model.py

```diff
@@ -64,13 +64,21 @@ def forward(self, x):
             y = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)
         else:
             # manual implementation of attention
-            att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
-            att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))
+            scale_factor = 1 / math.sqrt(k.size(-1)) 
+            att = (q @ k.transpose(-2, -1)) * scale_factor
+
+            # causal means the attention per token is not affected by future tokens, achieved by adding a bias
+            attn_bias = torch.zeros(k.size(-2), k.size(-2), dtype=q.dtype)
+            temp_mask = torch.ones(k.size(-2), k.size(-2), dtype=torch.bool).tril()
+            attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
+            att += attn_bias
+
             att = F.softmax(att, dim=-1)
-            att = self.attn_dropout(att)
+            att = self.attn_dropout(att) if self.training else att
```

Comment:
I believe this shouldn't be necessary since `nn.Dropout` will not apply dropout if `self.training = True`.

Dropout source: https://pytorch.org/docs/stable/_modules/torch/nn/modules/dropout.html#Dropout

Which calls `F.dropout`:

https://pytorch.org/docs/stable/generated/torch.nn.functional.dropout.html

Notice it takes a `training` arg which comes from `nn.Dropout`'s `self.training` state.